runtime/near-vm-runner/RUNTIMES.md

```diff
@@ -0,0 +1,176 @@
+for it to be considered correct. Any deviations from the specification would render an
+implementation incorrect.
+
+In addition to this, the NEAR protocol adds a requirement that all executions of a Wasm program
+must be deterministic. In other words, it must not be possible to observe different execution
+results based on the environment within which the VM runs.
+
+The Wasm specification specifies fast, safe and portable semantics to be one of its design goals.
+Even though Wasm specification does go to great lengths to make it possible to enable deterministic
+and reproducible execution, it is not one of the foundational goals. As thus, there are some parts
+of the specification that allow for non-deterministic, non-reproducible execution.
+
+A particularly troublesome area of the specification in this regard relates to the floating point
+support. The specification permits floating point operations to return a `NaN` result which the
+specification also permits to have multiple different bit representations and goes out of its way
+to specify the sign of the `NaN` result to be non-deterministic. The determinism and
+reproducibility requirements of a NEAR validator make it necessary for the runtime to ensure NaN
+signs and payloads are always deterministic.
+
+Another detail to keep an eye out is whether the floating point results are being rounded
+correctly. It is well known that the x87 co-processor does not necessarily round the results
+correctly, so a special attention to how a runtime implements the floating point operations on the
+32-bit x86 targets is warranted.
+
+## Reliability
+
+The required and desirable properties of a runtime implementation must remain true while the
+changes are made to it. Compilers and, by extension, runtimes are complex systems and a poor
+architecture may result in even simple changes introducing correctness or security bugs, which
+aren't necessarily easily discovered.
+
+A reliable code base in this sense should be architected in such a way that the algorithms they
+implement and the way they execute Wasm code remains correct by construction and is easy to verify.
+
+Size of a test suite, architecture of the code base and the number of developers/users are all
+reasonably correlated signals as to the reliability which one can expect from a runtime.
+
+## Platform support
+
+The NEAR validator implementation currently targets `x86_64-linux`, so stellar support for this
+target is a hard requirement.
+
+However, developers may not necessarily have a `x86_64` machine to develop with, or they might be
+using a device running a different operating system. With the recent surge in popularity of devices
+based on the Arm CPU architecture chips and generally high popularity of the macOS and Windows
+operating systems it would be a huge boon to the development experience if the runtime used also
+supported these other targets.
+
+## Runtime performance
```

Comment:
I think there is another important point to be mentioned about performance: Predictability

Essentially, our blockchain is a real-time system, as a transaction that consumes e.g. 1TGas should finish within 1ms. Optimizing performance for a RT system is a completely different engineering problem than optimization otherwise, since in RT, it is always the worst-case that counts.

This might sound like something that's only relevant to the preparation cost. But I would argue it is just as important for the generated code. For example, an optimization which tries to be smart with arrangement of more/less likely branches could be problematic for us, for several reasons that pop to my mind:
1) It doesn't improve worst-case performance but creates extra compiler overhead.
2) Gaining some extra performance on average doesn't help us a bit in terms of overall system performance, since we have to be conservative in how many transactions go in a block anyway.
3) It makes it harder for us to estimate a realistic worst-case execution time. As the average time per branch goes down, we are even tempted to lower the gas fee parameters for it. Which can lead to subtle security vulnerabilities which are potentially hard to spot.